cloudevents/extensions/opcua.md

```diff
@@ -0,0 +1,138 @@
+### source
+
+MUST either map to [Application
+Description](https://reference.opcfoundation.org/Core/Part4/v104/docs/7.1) field
+`applicationUri` of the OPC UA server or to a customer configured identifier
+like a unified namespace path.
+
+### type
+
+MUST map to [Data Set Message Header](https://reference.opcfoundation.org/Core/Part14/v105/docs/7.2.5.4#Table164) field `MessageType`.
+
+## Mapping of OPTIONAL Attributes
+
+### datacontenttype
+
+MUST be `application/json` for OPC UA PubSub JSON payload and MAY be extended by
+`+gzip` when payload is gzip compressed.
+
+### dataschema
+
+OPC UA provides type information as part of PubSub metadata messages, for non
+OPC UA consumers or when different payload encoding like Avro is used, it is
+REQUIRED to provide schema information (based on metadata information) in a
+separate format like [JSON schema](https://json-schema.org/specification) or
+[Avro schema](https://avro.apache.org/docs/1.11.1/specification/) or others. For
+those cases the Attribute references the schema and is used for versioning.
+
+### subject
+
+For metadata and data messages (type one of `ua-metadata`, `ua-keyframe`,
+`ua-deltaframe`), `subject` MUST map to either [Data Set Message
+Header](https://reference.opcfoundation.org/Core/Part14/v105/docs/7.2.5.4#Table164)
+field `DataSetWriterId` or `DataSetWriterName`.
+
+For event messages (type equals to `ua-event`) `subject` MUST map to [Base Event
+Type](https://reference.opcfoundation.org/Core/Part5/v104/docs/6.4.2) field
+`EventId`. 
+
+### time
+
+MUST map to [Data Set Message
+Header](https://reference.opcfoundation.org/Core/Part14/v105/docs/7.2.5.4#Table164)
+field `Timestamp`.
+
+## Mapping for other extensions
+
+The following well-known extensions attributes are used for data messages and
+event messages (type one of `ua-keyframe`, `ua-deltaframe`, `ua-event`).
+
+### sequence
+
+Attribute as defined by [sequence extensions](./sequence.md) MUST map to [Data
+Set Message Header](https://reference.opcfoundation.org/Core/Part14/v105/docs/7.2.5.4#Table164)
+field `SequenceNumber`.
+
+### traceparent
+
+Attribute as defined by [distributed-tracing extension](./distributed-tracing.md)
+MUST be used to allow tracing from event publisher towards consumer.
+
+### tracestate
+
+Attribute as defined by [distributed-tracing extension](./distributed-tracing.md)
+MAY be used to allow tracing from event publisher towards consumer.
+
+### recordedtime
+
+Attribute as defined by [recordedtime extension](./recordedtime.md) MUST be used
+to determine the latency between event publisher towards consumer. 
+
+## Attributes
+
+### opcuametadatamajorversion
+
+- Type: `Integer`
+- Description: Links dataset message to the current version of the metadata.
+Contains value from `MajorVersion` of [Data Set Message Header](https://reference.opcfoundation.org/Core/Part14/v105/docs/7.2.5.4#Table164) field `MetaDataVersion`. 
+- Constraints
+  - OPTIONAL
+  - MAY be omitted if `dataschema` is used
```

Comment:
As written, we don't need both of these. What is the intent here? REQUIRED unless `dataschema` is used? Or 100% OPTIONAL all the time? or SHOULD be used in general but SHOULD NOT be used when `dataschema` is present?

I have the same questions about the rest of the attributes.

Reply:
As per OPC UA PubSub spec the `MetaDataVersion` is a structed defined out of 2 elements `MinorVersion` and `MajorVersion`. I avoided adding an JSON object to the header, and splitted it into 2 attributes.

The `MetaDataVersion` is per OPC UA PubSub spec optional (all the time) _but_ this is the field to use, to map between an metadata message version and a data/event message.

`opcuastatus` is optional as long as there is no error while reading the data/events. If there is an error is must be present, if there is no error it may be omitted.

Reply:
.. and I missed, if the major version is updated, the minor needs to have the same value and after that if there are only compatible additions (new elements added at the end of the payload) only the minor version will be updated. 

But if `dataschema` is used, the versioning would be done via that attribute.

Reply:
It sounds like it should be:
- OPTIONAL but MUST NOT be present if `dataschema` is used

no?

Not following the `opcustatus` comment though. Do you mean:
- REQUIRED if the event is related to an error, otherwise it is OPTIONAL

?

Reply:
I used your proposal for metadataversion 👍 

Regarding opcuastatus, in OPC UA PubSub will a client subscribe to value changes or events. There are typically multiple OPC UA nodes subscribed at the same time. The publishing of any node might fail, and the status is the collective status of all subscribed nodes. If everything is good, it can be omitted if one or more have error it must be present